File: tests/golden/02-interfaces-classes.ts
/**
 * 測試 02: 介面與類別
 */

// 基本介面
interface User {
  id: string;
  name: string;
  email?: string;
  readonly createdAt: Date;
}

// 介面繼承
interface Admin extends User {
  role: 'admin';
  permissions: string[];
}

// 索引簽名
interface Dictionary {
  [key: string]: any;
}

// 類別實作介面
class UserImpl implements User {
  readonly createdAt: Date;

  constructor(
    public id: string,
    public name: string,
    public email?: string
  ) {
    this.createdAt = new Date();
  }

  greet(): string {
    return `Hello, I'm ${this.name}`;
  }
}

// 類別繼承
class AdminUser extends UserImpl implements Admin {
  role: 'admin' = 'admin';

  constructor(
    id: string,
    name: string,
    email: string,
    public permissions: string[]
  ) {
    super(id, name, email);
  }

  hasPermission(permission: string): boolean {
    return this.permissions.includes(permission);
  }
}

// 靜態成員
class Counter {
  private static instance: Counter;
  private count: number = 0;

  static getInstance(): Counter {
    if (!Counter.instance) {
      Counter.instance = new Counter();
    }
    return Counter.instance;
  }

  increment(): number {
    return ++this.count;
  }

  getCount(): number {
    return this.count;
  }
}

export { User, Admin, UserImpl, AdminUser, Counter };

File: tests/golden/expected/02-interfaces-classes.go
package main

import (
	"fmt"
	"time"
)

type User struct {
	Id        string
	Name      string
	Email     *string
	CreatedAt time.Time
}

type Admin struct {
	User
	Role        string
	Permissions []string
}

type Dictionary map[string]interface{}

type UserImpl struct {
	Id        string
	Name      string
	Email     *string
	CreatedAt time.Time
}

func NewUserImpl(id string, name string, email *string) *UserImpl {
	return &UserImpl{
		Id:        id,
		Name:      name,
		Email:     email,
		CreatedAt: time.Now(),
	}
}

func (u *UserImpl) Greet() string {
	return fmt.Sprintf("Hello, I'm %s", u.Name)
}

type AdminUser struct {
	UserImpl
	Role        string
	Permissions []string
}

func NewAdminUser(id string, name string, email string, permissions []string) *AdminUser {
	emailPtr := &email
	return &AdminUser{
		UserImpl:    *NewUserImpl(id, name, emailPtr),
		Role:        "admin",
		Permissions: permissions,
	}
}

func (a *AdminUser) HasPermission(permission string) bool {
	for _, p := range a.Permissions {
		if p == permission {
			return true
		}
	}
	return false
}

type Counter struct {
	count int
}

var counterInstance *Counter

func GetCounterInstance() *Counter {
	if counterInstance == nil {
		counterInstance = &Counter{count: 0}
	}
	return counterInstance
}

func (c *Counter) Increment() int {
	c.count++
	return c.count
}

func (c *Counter) GetCount() int {
	return c.count
}